src/ComfortPay/Message/ChargeRequest.php

```diff
@@ -18,79 +24,163 @@ public function setTerminalId($value)
         return $this->setParameter('terminalId', $value);
     }
 
-    public function getWs()
+    public function getParentTransactionId()
+    {
+        return $this->getParameter('parentTransactionId');
+    }
+
+    public function setParentTransactionId($value)
+    {
+        return $this->setParameter('parentTransactionId', $value);
+    }
+
+    public function getReferedCardId()
+    {
+        return $this->getParameter('referedCardId');
+    }
+
+    public function setReferedCardId($value)
+    {
+        return $this->setParameter('referedCardId', $value);
+    }
+
+    public function getIpspData()
+    {
+        return $this->getParameter('ipspData');
+    }
+
+    public function setIpspData($value)
     {
-        return $this->getParameter('ws');
+        return $this->setParameter('ipspData', $value);
     }
 
-    public function setWs($value)
+    public function getE2eReference()
     {
-        return $this->setParameter('ws', $value);
+        return $this->getParameter('e2eReference');
     }
 
-    public function getCid()
+    public function setE2eReference($value)
     {
-        return $this->getParameter('cid');
+        return $this->setParameter('e2eReference', $value);
     }
 
-    public function setCid($value)
+    public function getSubmerchantId()
     {
-        return $this->setParameter('cid', $value);
+        return $this->getParameter('submerchantId');
     }
 
-    public function getTransactionId()
+    public function setSubmerchantId($value)
     {
-        return $this->getParameter('transactionId');
+        return $this->setParameter('submerchantId', $value);
     }
 
-    public function setTransactionId($value)
+    public function getLocation()
     {
-        return $this->setParameter('transactionId', $value);
+        return $this->getParameter('location');
+    }
+
+    public function setLocation($value)
+    {
+        return $this->setParameter('location', $value);
+    }
+
+    public function getCity()
+    {
+        return $this->getParameter('city');
+    }
+
+    public function setCity($value)
+    {
+        return $this->setParameter('city', $value);
+    }
+
+    public function getAlpha2CountryCode()
+    {
+        return $this->getParameter('alpha2CountryCode');
+    }
+
+    public function setAlpha2CountryCode($value)
+    {
+        return $this->setParameter('alpha2CountryCode', $value);
     }
 
     public function getData()
     {
-        $this->validate('terminalId', 'ws', 'amount', 'currency', 'transactionId', 'cid');
+        $this->validate('terminalId', 'amount', 'transactionId', 'transactionType', 'referedCardId', 'mid', 'currency');
```

Comment:
You should also validate `parentTransactionId` if the `transactionType` is one of these: `PREAUTH-CONFIRM`, `PREAUTH-CANCEL`, `CHARGEBACK`

(based on the docs)